pkg/solana/chainreader/event_read_binding.go

```diff
@@ -0,0 +1,229 @@
+	namespace, genericName string,
+	indexedSubkeys map[string]string,
+	events EventsReader,
+) *eventReadBinding {
+	binding := &eventReadBinding{
+		namespace:      namespace,
+		genericName:    genericName,
+		indexedSubkeys: indexedSubkeys,
+		events:         events,
+	}
+
+	binding.remapper = remapHelper{binding.remapPrimitive}
+
+	return binding
+}
+
+func (b *eventReadBinding) SetAddress(key solana.PublicKey) {
+	b.key = key
+}
+
+func (b *eventReadBinding) GetAddress(_ context.Context, _ any) (solana.PublicKey, error) {
+	return b.key, nil
+}
+
+func (b *eventReadBinding) SetCodec(codec types.RemoteCodec) {
+	b.codec = codec
+}
+
+func (b *eventReadBinding) SetModifier(modifier commoncodec.Modifier) {
+	b.modifier = modifier
+}
+
+func (b *eventReadBinding) CreateType(forEncoding bool) (any, error) {
+	itemType := codec.WrapItemType(forEncoding, b.namespace, b.genericName, codec.ChainConfigTypeEventDef)
+
+	return b.codec.CreateType(itemType, forEncoding)
+}
+
+func (b *eventReadBinding) Decode(ctx context.Context, bts []byte, outVal any) error {
+	itemType := codec.WrapItemType(false, b.namespace, b.genericName, codec.ChainConfigTypeEventDef)
+
+	return b.codec.Decode(ctx, bts, outVal, itemType)
+}
+
+func (b *eventReadBinding) QueryKey(
+	ctx context.Context,
+	filter query.KeyFilter,
+	limitAndSort query.LimitAndSort,
+	sequenceDataType any,
+) ([]types.Sequence, error) {
+	var err error
+
+	if filter, err = b.remapper.remap(filter); err != nil {
+		return nil, err
+	}
+
+	itemType := strings.Join([]string{b.namespace, b.genericName}, ".")
+
+	logs, err := b.events.FilteredLogs(ctx, filter.Expressions, limitAndSort, itemType)
+	if err != nil {
+		return nil, err
+	}
+
+	sequences, err := b.decodeLogsIntoSequences(ctx, logs, sequenceDataType)
+	if err != nil {
+		return nil, err
+	}
+
+	return sequences, nil
+}
+
+func (b *eventReadBinding) remapPrimitive(expression query.Expression) (query.Expression, error) {
+	var (
+		comp query.Expression
+		err  error
+	)
+
+	switch primitive := expression.Primitive.(type) {
+	case *primitives.Comparator:
+		if comp, err = b.encodeComparator(primitive); err != nil {
+			return query.Expression{}, fmt.Errorf("failed to encode comparator %q: %w", primitive.Name, err)
+		}
+	case *primitives.Confidence:
+		// TODO: maybe use an ignore filter?
+		// confidence is ignored in solana
+	}
+
+	return comp, nil
+}
+
+func (b *eventReadBinding) encodeComparator(comparator *primitives.Comparator) (query.Expression, error) {
+	onChainName, ok := b.indexedSubkeys[comparator.Name]
+	if !ok {
+		return query.Expression{}, fmt.Errorf("%w: unknown indexed subkey mapping %s", types.ErrInvalidConfig, comparator.Name)
+	}
+
+	itemType := strings.Join([]string{b.namespace, b.genericName, comparator.Name}, ".")
+
+	for idx, comp := range comparator.ValueComparators {
+		// need to do a transform and then extract the value for the subkey
+		newValue, err := b.modifier.TransformToOnChain(comp.Value, itemType)
```

Comment:
We can't call this function, unless we modify the codec to skip the Borsh-encoding step.

I'm still not clear on whether we actually need to call this at all, but if we do then it either has to be without the Borsh-encoding or we'll need to decode it as soon as it finishes encoding.

Reply:
`TransformToOnchain` is a modifier function and doesn't handle encoding/decoding.